### PR TITLE
[dualtor] Update mux simulator control utility

### PR DESCRIPTION
**Pull request by bingwang-ms**
Signed-off-by: bingwang <bingwang@microsoft.com>

### Description of PR

Summary:
Fixes # (issue)
This PR is to update mux simulator control utility.

1. Refactor helper function into function level fixture
2. Change argument physical_port to interface_name
3. Add a new fixture simulator_server_down to simulate server down scenario.

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
This PR is to update mux simulator control utility.

#### How did you do it?
Please refer to code.

#### How did you verify/test it?
Verified with a demo script.
```
def test_demo_1(toggle_simulator_port_to_upper_tor, toggle_simulator_port_to_lower_tor, check_simulator_read_side, upper_tor_host, lower_tor_host, get_active_torhost):
    toggle_simulator_port_to_lower_tor("Ethernet0")
    assert check_simulator_read_side("Ethernet0") == 2
    assert get_active_torhost("Ethernet0") == lower_tor_host
 
    toggle_simulator_port_to_upper_tor("Ethernet0")
    assert check_simulator_read_side("Ethernet0") == 1
    assert get_active_torhost("Ethernet0") == upper_tor_host
    assert True
 
def test_demo_2(simulator_server_down, recover_all_directions):
    simulator_server_down("Ethernet0")
    recover_all_directions("Ethernet0")
```

#### Any platform specific information?
Dualtor testbed specific.

#### Supported testbed topology if it's a new test case?
No.

### CODE REVIEW

File: tests/common/dualtor/mux_simulator_control.py [hunk 1]
```diff
@@ -36,23 +36,31 @@ def mux_server_url(request, tbinfo):
     port = utilities.get_group_visible_vars(inv_files, server, 'mux_simulator_port')
     return "http://{}:{}/mux/{}".format(ip, port, vmset_name)
 
-def _url(server_url, physical_port=None, action=None):
+@pytest.fixture
+def url(mux_server_url, duthost, tbinfo):
```
Comment (neethajohn) [inline tests/common/dualtor/mux_simulator_control.py:40]: Is tbinfo needed?

File: tests/common/dualtor/mux_simulator_control.py [hunk 2]
```diff
@@ -107,114 +115,145 @@ def _post(server_url, data):
...
-    Function to toggle a given y_cable ports to lower_tor
-    Args:
-        mux_server_url: a str, the address of mux server
-        physical_port: physical port on switch, an integer starting from 1
-    """
-    _simulator_port_toggle_to(mux_server_url, physical_port, LOWER_TOR)
+@pytest.fixture
+def toggle_simulator_port_to_upper_tor(url):
+    """
+    Returns _toggle_simulator_port_to_upper_tor to make fixture accpet arguments
```
Comment (neethajohn) [inline tests/common/dualtor/mux_simulator_control.py:161]: spell check 'accept'

File: tests/common/dualtor/mux_simulator_control.py [hunk 3]
```diff
@@ -107,114 +115,145 @@ def _post(server_url, data):
...
-def toggle_simulator_port_to_lower_tor(mux_server_url, physical_port):
-    """
-    Function to toggle a given y_cable ports to lower_tor
-    Args:
-        mux_server_url: a str, the address of mux server
-        physical_port: physical port on switch, an integer starting from 1
-    """
-    _simulator_port_toggle_to(mux_server_url, physical_port, LOWER_TOR)
+@pytest.fixture
+def toggle_simulator_port_to_upper_tor(url):
```
Comment (neethajohn) [inline tests/common/dualtor/mux_simulator_control.py:159]: Is there any specific reason why there are diff fixtures for toggle to upper and lower? Can these be combined to toggle_simulator_port_to_tor(url, target_tor)

File: tests/common/dualtor/mux_simulator_control.py [hunk 4]
```diff
@@ -107,114 +115,145 @@ def _post(server_url, data):
...
-        physical_port: physical port on switch, an integer starting from 1
-        directions: a list, may contain "upper_tor", "lower_tor", "nic"
-    Returns:
-        None.
-    """
-    server_url = _url(mux_server_url, physical_port, DROP)
-    data = {"out_ports": directions}
-    pytest_assert(_post(server_url, data), "Failed to set drop on {}".format(directions))
+@pytest.fixture
+def set_drop(url):
```
Comment (neethajohn) [inline tests/common/dualtor/mux_simulator_control.py:119]: Can both the set_drop and set_output be combined to a single fixture?

File: tests/common/dualtor/mux_simulator_control.py [hunk 5]
```diff
@@ -107,114 +115,145 @@ def _post(server_url, data):
...
+    """
+    def _check_simulator_read_side(interface_name):
+        """
+        Retrieve the current active tor from y_cable simulator server.
+        Args:
+            interface_name: a str, the name of interface to control
+        Returns:
+            1 if upper_tor is active
+            2 if lower_tor is active
+            -1 for exception or inconstient status
```
Comment (neethajohn) [inline tests/common/dualtor/mux_simulator_control.py:225]: spell check 'inconsistent'

File: tests/common/dualtor/mux_simulator_control.py (hunk 1 above)
Comment (bingwang-ms) [reply tests/common/dualtor/mux_simulator_control.py]: Good catch. Updated. Thanks!

File: tests/common/dualtor/mux_simulator_control.py (hunk 2 above)
Comment (bingwang-ms) [reply tests/common/dualtor/mux_simulator_control.py]: Updated. Thanks!

File: tests/common/dualtor/mux_simulator_control.py (hunk 3 above)
Comment (bingwang-ms) [reply tests/common/dualtor/mux_simulator_control.py:159]: Yes. It's possible to merge ```toggle_simulator_port_to_upper_tor``` and ```toggle_simulator_port_to_lower_tor``` into a common fixture ```toggle_simulator_port_to_tor```, but that's means the caller have to know more internal details and pass correct arguments to the interface, such as ```lower_tor``` or ```upper_tor```. Spliting them into two separated call would be more friendly to caller, I think.

File: tests/common/dualtor/mux_simulator_control.py (hunk 4 above)
Comment (bingwang-ms) [reply tests/common/dualtor/mux_simulator_control.py:119]: Same reason as above. I'd like to split them into two function in order to be more friendly to caller.

File: tests/common/dualtor/mux_simulator_control.py (hunk 5 above)
Comment (bingwang-ms) [reply tests/common/dualtor/mux_simulator_control.py]: Thanks. Updated.

**Review by neethajohn — APPROVED** [review]

**Pull request merged**